Added ability to remove project members

Slartibartfass2 opened the pull request:
Closes #19 

## What I have made

- I noticed that the tab list of the navigation bar wraps its tabs (buttons) in an \<a\> tag, therefore the button is never actually used (click events are handled by the \<a\> tag)
  - I changed the tabs trigger to directly be an \<a\> tag when `href` is provided
  - the change can be seen in https://github.com/SE-UUlm/snowballr-frontend/pull/191/commits/24f28708338b2acbac53e913c29b214eff4c807e
- added a reusable alert dialog, which is used in the create project dialog and as remove project member dialog
- added the remove project member dialog
- added e2e tests for invite users workflow
- added e2e tests for remove project members workflow

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - ~~[ ] unit tests~~ (no logic added that needs unit tests)
  - [x] integration tests
  - [x] end-to-end tests
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 committed:
feat: add reusable alert dialog

Slartibartfass2 committed:
feat: use reusable alert dialog for create project dialog

Slartibartfass2 committed:
feat: add remove member dialog

Slartibartfass2 committed:
feat: disable send invitations button when input is empty

Slartibartfass2 committed:
feat: filter out already invited members

Slartibartfass2 committed:
test: adjust e2e tests to new dialogs

Slartibartfass2 committed:
test: use user fixture for later use

Slartibartfass2 committed:
test: add invite users tests

Slartibartfass2 committed:
refactor: make tab list trigger a tag when href is provided

Previously, we an a tag with a nested button, but the button is never clicked.
Changing the logic of the tabs trigger so that it switches to an a tag when href is provided
fixes this as now only the a tag is shown, but when no href is provided a button is rendered instead.

Slartibartfass2 committed:
test: add e2e tests for remove project members

Slartibartfass2 committed:
test: increase timeout for test with many operations

mowi12 commented on src/lib/components/composites/dialog/AlertDialog.svelte:
````diff
@@ -0,0 +1,142 @@
+    </AlertDialog>
+```
+
+Usage:
````
Isn't this usage useless if you provide an example before?

Slartibartfass2 replied:
I see your point that it feels duplicated. The first example is about binding the properties and the second one is about the general usage. I think more documentation is better than less, even if there's a bit duplicated.

mowi12 replied:
Probably true. Then we keep it.

mowi12 commented on src/lib/components/primitives/alert-dialog/alert-dialog-cancel.svelte:
Why did you change the type? This doesn't change anything.

Slartibartfass2 replied:
Not really no, as we do not change the type. I mainly did it for exporting, but I realized that I could've used the type directly. However, this appears more clean to me.

mowi12 replied:
I understand. But if you want to keep it clean and consistent don't you need to update all the files, e.g. `alert-dialog-description.svelte`. Else, I don't see the reason to keep this modification.

Slartibartfass2 replied:
But we don't export these props. We only do this for the action and cancel button, which is why I made the modification there.

mowi12 replied:
Well, I guess you're right. Since these are no major files and this isn't really important, I think we can agree to keep the current state.

mowi12 commented on playwright.config.ts:
```diff
@@ -4,18 +4,22 @@ import type { TestOptions } from "./tests/e2e/fixtures/general-fixture";
         ["html", { outputFolder: "e2e-report" }],
     ],
+    expect: {
+        timeout: 2_500,
```
2.5 seconds seem a bit short. Did you test it with shorter times, and they worked as well, and 2.5 seconds is already with puffer?

Slartibartfass2 replied:
I checked this again and the longest I've got was 1.3s, so I think 2.5s is fine. Initially, I chose this because it's half of the default.

As I use a slow ass machine, this should be fine. Did you experience any errors with that?

mowi12 replied:
No, I experienced no errors, I just wanted to make sure you tested it and didn't just choose some value.

mowi12 commented on tests/integration/navigation-bar/navigation-bar.test.ts:
```diff
@@ -105,7 +104,7 @@ describe("NavigationBar", () => {
         });
 
-        const linkTags = screen.getAllByRole("link");
+        const linkTags = screen.queryAllByRole("link");
```
What's the difference between `get` and `query`?

Slartibartfass2 replied:
`get` throws if there's nothing, `query` just returns a list, which can also be empty.
Now that the `<a>` tags have the `tab` role, they don't have the `link` role anymore, which is why this wouldn't have worked with `get`. While there are many ways to test this, this was the one way with the least difference.

mowi12 replied:
Makes sense.

mowi12 commented on tests/integration/settings/project-settings/review/keyword-settings.test.ts:
```diff
@@ -84,5 +84,5 @@ describe("KeywordSettings", () => {
         expect(screen.queryByText("Maximum number of keywords reached!")).toBeInTheDocument();
         expect(screen.queryByText(`New Tag ${maximumAmountOfTags}`)).not.toBeInTheDocument();
-    });
+    }, 10000);
```
Does this actually reduce the amount of flaky tests you receive?

Slartibartfass2 replied:
Yes, the test timed out on some occasions, which is why increased the timeout. I added a comment for clarification.

mowi12 replied:
Okay, thanks!

mowi12 commented on src/lib/components/composites/settings/project-settings/members/RemoveMemberDialog.svelte:
```diff
@@ -0,0 +1,79 @@
+    {/snippet}
+    {#snippet description()}
+        Once removed, <span class="font-bold">{memberName}</span> will no longer have access to this
+        project. You can always re-invite them later.
```
What about the users reviews?
Should this also be stated in here?

Slartibartfass2 replied:
Do we have a defined procedure for that?

mowi12 replied:
If we have one, you should know it, as it would have been decided early on in the project, or not?
But I would be fine if we omit this information or just add a comment with 'TODO: Clarify what happens with deleted users reviews'. What do you think?

Slartibartfass2 replied:
Sounds great!

mowi12 commented on src/lib/components/primitives/button/button.svelte:
```diff
@@ -9,6 +9,8 @@
             variant: {
                 default: "bg-primary text-primary-foreground hover:bg-primary/90",
                 destructive: "bg-destructive text-destructive-foreground hover:bg-destructive/90",
+                "destructive-subtle":
```
```suggestion
                destructiveSubtle:
```

mowi12 replied:
Don't forget to update all occurences.

Slartibartfass2 replied:
Done.

mowi12 commented on src/routes/project/[projectId]/settings/members/+page.svelte:
```diff
@@ -27,27 +28,43 @@
         // First fetch the members again and only then replace them, so that no loading state is shown
         await loadMembers({ id: projectId })
             .then((members) => {
                 loadingMembersLocal = Promise.resolve(members);
```
As i should have already commented in the main review comment, I think you need to invalidate something in order to be able to invite the last removed user directly without refreshing the site.

Slartibartfass2 replied:
I fixed it. The only problem was that in `InviteUsersDialog` the list of members is not updated.

Slartibartfass2 replied:
I added a test that ensures that this works.

mowi12 replied:
Perfect!

mowi12 commented on tests/integration/settings/project-settings/members/remove-member-dialog.test.ts:
```diff
@@ -0,0 +1,138 @@
+        );
+    });
+
+    test("When the member is an admin, then the trigger is disabled", async () => {
```
I assume you mean the member to remove?

Slartibartfass2 replied:
Yes, that's correct. But this what the component is about. Do you think we require a more detailed description?

mowi12 replied:
No, you're right. It now makes sense. It was getting pretty late as I reviewed this.

mowi12 commented on tests/integration/settings/project-settings/members/remove-member-dialog.test.ts:
```diff
@@ -0,0 +1,138 @@
+        expect(trigger).toBeDisabled();
+    });
+
+    test("When the member is the current user, then the trigger is disabled", async () => {
```
Same here.

Slartibartfass2 replied:
See above.

mowi12 commented on tests/e2e/fixtures/general-fixture.ts:
```diff
@@ -60,4 +63,19 @@ export const mockBackendServiceTest = base.extend<
+ * This test user is used is registered before each test.
+ */
+export const userTest = base.extend<{ user: TestUser }>({
+    user: [testUser, { option: true }],
```
What is the `option` field doing?

Slartibartfass2 replied:
`option` makes the parameter available in `playwright-config`. But as we do not use this, this can be removed. Good catch!

mowi12 commented on tests/e2e/fixtures/project-members-settings-fixtures.ts:
```diff
@@ -0,0 +1,25 @@
+
+        await use(projectMembersSettingsPage);
+
+        await mockBackendService.softDeleteProject({ id: project.id });
```
Why do you already delete this project here?

Slartibartfass2 replied:
This is after the test is executed.
`use` executes the test, i.e., the fixture is a wrapper of the test.
Everything before `use` is the before each hook, and everything after `use` is the after each hook.

mowi12 replied:
Ahh, good to know.

mowi12 commented on tests/e2e/helper.ts:
```diff
@@ -14,3 +14,25 @@ export async function reloadWait(page: Page, element: Locator) {
+ * @returns the first and last name of the user
+ */
+export async function getCurrentUser(page: Page): Promise<{ firstName: string; lastName: string }> {
+    await page.getByRole("button", { name: /[A-Z]{2}/ }).click();
```
```suggestion
    await page.getByRole("button", { name: /^[A-Z]{2}$/ }).click();
```

If there are any two capital letters in the project which are on a button, they would also match, now they need to be at the start and end of a line making it more rare to match.
-> `home-page-model.ts`

Slartibartfass2 replied:
Good catch

mowi12 commented on tests/e2e/helper.ts:
```diff
@@ -14,3 +14,25 @@ export async function reloadWait(page: Page, element: Locator) {
+ * @param page - the current page
+ * @returns the first and last name of the user
+ */
+export async function getCurrentUser(page: Page): Promise<{ firstName: string; lastName: string }> {
```
Imo this function name is a little bit misleading, isn't it?

Slartibartfass2 replied:
What is misleading you?

mowi12 replied:
I mean, you kind of get the current user but only the name. If someone else wants to write a test and sees this, wouldn't he assume he gets the full user object? Yes, the documentation states that this only returns the first and last name.
But as I do not have a better solution, I think we can keep this as is.

Resolve on your own if you also want to keep it.

Slartibartfass2 replied:
I renamed it.

mowi12 commented on tests/e2e/invite-users.test.ts:
```diff
@@ -0,0 +1,112 @@
+import { getCurrentUser } from "./helper";
+
+test.describe("Inviting users to a project", () => {
+    /**
+     * Here we test for project members. In the fixture we create a project with one admin and one invitee.
+     */
```
Is there a reason why this comment is inside `test.describe` and not above?

Slartibartfass2 replied:
No, not really. Seemed like the best place to me.

mowi12 replied:
I guess we can keep it here as it is no documentation of the method, you're right.

mowi12 commented on tests/e2e/pom/project-member-settings-page-model.ts:
```diff
@@ -0,0 +1,135 @@
+
+    /**
+     * Checks whether the user with the passed userId is listed in the project members.
+     */
```
Missing `@param`.

Slartibartfass2 replied:
Great catch. I added more documentation.

mowi12 reviewed (changes requested):
Wow, this looks really good. I don't see any logical error or similar. I only have some minor adjustments that we need to talk about and two problems I faced.

1. After removing an user from the project, this user is not in the list of users you can add when you click on the 'Invite Users' button. Only after reloading the page, this users is in this list.

2. <img width="551" alt="Screenshot 2025-05-14 at 17 15 08" src="https://github.com/user-attachments/assets/492d5372-9ee8-492e-bf55-954a90e7e423" />

This happens when the user name is the email adress.

Slartibartfass2 committed:
fix: update possible invitees list after member is removed

Slartibartfass2 committed:
refactor: make pluralize directly accept arrays too

Slartibartfass2 committed:
style: rename 'destructive-subtle' to 'destructiveSubtle'

Slartibartfass2 committed:
chore: add minor review changes

Slartibartfass2 committed:
test: add test for reinviting removed user

Slartibartfass2 committed:
feat: remove actual member name from remove member dialog action button

For long names this would've caused the dialog to overflow. As the name is
shown in the title and the description, we don't require it to be on the
action button as well.

Slartibartfass2 reviewed (commented):
Thank you very much for the fast review of these many changes.

Next to changing what you've noticed, I also changed the signature of `pluralize` to also accept an array, as this is used frequently. Additionally, I removed the member name from the action button of the dialog as this often caused an overflow.

Slartibartfass2 committed:
fix: fix loading members functionality

mowi12 reviewed (changes requested):
Nice, you fixed all the larger problems. There are only some minor discussions we need to finish, but this should handled very fast!

mowi12 reviewed (changes requested):
I know, no one has an e-mail adress this long, but my own e-mail adress was pretty close. Is there any way to handle this problem? I don't have any good idea how do handle this case. Splitting the address feels wrong, but may be the only solution.

But in every case, the buttons should not shift that much to the right.

<img width="624" alt="Screenshot 2025-05-15 at 17 26 35" src="https://github.com/user-attachments/assets/59fd2a52-dac4-4892-8062-0faa2db2a9ad" />

Slartibartfass2 committed:
refactor: renamed 'getCurrentUser' to ' getNameOfCurrentUser'

Slartibartfass2 committed:
feat: wrap long emails

Slartibartfass2 reviewed (commented):
Thank you for the fast feedback.

I found a way to wrap long emails, what are your thoughts on this?

mowi12 reviewed (changes requested):
Nicely done! Looks great. Now only one open conversation is left. Then we are good to go!

mowi12 reviewed (dismissed):
Nice. We really got some work done.

While testing, I found a [bug](https://github.com/SE-UUlm/snowballr-frontend/issues/377), but I don't think this is part of this issue, since it has nothing to do with your logic.

Slartibartfass2 committed:
chore: pin mock backend version for e2e tests

Currently, the CI fails because of the latest changes in the mock backend
